feat: add `list_price` to response of can_redeem

pwnage101 opened the pull request:
This also bundles other improvements:
* output serializers are defined for the can_redeem endpoint, enabling much higher quality redoc pages
* added a ContentKeyField to replace the ValidateContentKeyMixin which was harder to use and overly complicated.

ENT-7137

pwnage101 commented on enterprise_access/apps/api/v1/views/subsidy_access_policy.py:
```diff
@@ -566,22 +591,41 @@ def can_redeem(self, request, enterprise_customer_uuid):
+            # Determine the price for content for display purposes only.
+            try:
+                content_metadata = self.subsidy_client.get_subsidy_content_data(enterprise_customer_uuid, content_key)
+                # Note that the "content_price" key is guaranteed to exist, but the value may be None.  Also, the
+                # upstream field type is a string representing a float, and we passthrough that verbatim.
+                content_price = content_metadata["content_price"]
```
FYI, the code as written just passes-through a string representing decimal dollars (which is how the list_price is stored in the content metadata upstream), rather than integer cents used in the rest of this service, including in other API responses.  My personal opinion would be to have this convert the output to be consistent with the rest of the quantities returned by this API, but I also was just referencing the API spec: https://github.com/openedx/enterprise-access/blob/6ad37aac9b3fbdf594c2409fd8b7ad8aa7ab3555/docs/decisions/0006-api-specification-for-enterprise-mfes.rst#L1

Open to suggestions!

pwnage101 replied:
FYI, I just looked at the API spec and discovered that integer cents was called for, not a passthrough string representing decimal dollars. https://github.com/openedx/enterprise-access/blob/6ad37aac9b3fbdf594c2409fd8b7ad8aa7ab3555/docs/decisions/0006-api-specification-for-enterprise-mfes.rst#L107

I plan to convert this before merging.

iloveagent57 commented on enterprise_access/apps/api/v1/views/subsidy_access_policy.py:
```diff
@@ -566,22 +591,41 @@ def can_redeem(self, request, enterprise_customer_uuid):
+            element_responses.append(element_response)
 
-        return Response(response, status=status.HTTP_200_OK)
+        response_serializer = serializers.SubsidyAccessPolicyCanRedeemResponseSerializer(
+            {"results": element_responses}
+        )
+        return Response(response_serializer.data, status=status.HTTP_200_OK)
```
Nice, way easier to reason about everything wrapped up in here now.

iloveagent57 reviewed (approved):
lgtm, I'll take another look once you do that stuff with `list_price`.

pwnage101 commented:
see 2nd commit for changes.  changed the list_price output to an integer, and also refactored the way of mocking the API to get around the lru_cache persisting between tests.

adamstankiewicz commented on enterprise_access/apps/api/v1/views/subsidy_access_policy.py:
```diff
@@ -566,22 +591,43 @@ def can_redeem(self, request, enterprise_customer_uuid):
 
-        return Response(response, status=status.HTTP_200_OK)
+        response_serializer = serializers.SubsidyAccessPolicyCanRedeemResponseSerializer(
+            {"results": element_responses}
```
This is technically a breaking change to update the response from a list to a dict with a `results` list, as the client (learner portal MFE) currently expects a list to be returned. If we change this, it will likely result in a JavaScript error when parsing its response in the learner portal course page on stage.

A couple links to learner portal source where we parse the `can-redeem` response:
* https://github.com/openedx/frontend-app-learner-portal-enterprise/blob/2e8ebf2beac823651019338352dd2841d7b860c2/src/components/course/data/hooks.jsx#L552-L557
* https://github.com/openedx/frontend-app-learner-portal-enterprise/blob/2e8ebf2beac823651019338352dd2841d7b860c2/src/components/course/data/hooks.jsx#L567

![image](https://github.com/openedx/enterprise-access/assets/2828721/2bec377b-5be2-4bd3-91ac-1ab9b4d2d74b)

A few things to consider: 
1. be OK with potentially breaking stage learner portal course page until it is updated to reflect this change shortly after (will not impact production).
2. version the endpoint.
3. update learner portal course page to handle both style of responses and then remove the old way.
4. something else?

pwnage101 replied:
Good catch, I forgot to mention this.  I re-worked the endpoint to revert it back to original behavior, with ample code comments to explain why we chose to deviate from the industry norm.  Also, I re-worked it in a way that the redoc page looks accurate:
![image](https://github.com/openedx/enterprise-access/assets/85151/a29bd560-e894-4c12-8a3a-b8eec961ef45)

adamstankiewicz commented on enterprise_access/apps/api/v1/views/subsidy_access_policy.py:
```diff
@@ -566,22 +591,43 @@ def can_redeem(self, request, enterprise_customer_uuid):
+                # Note that the "content_price" key is guaranteed to exist, but the value may be None.  Also, the
+                # upstream field type is a string representing a decimal dollar.
+                list_price_decimal_dollars_str = content_metadata["content_price"]
+                list_price_integer_cents = int(float(list_price_decimal_dollars_str) * 100)
```
[curious] Is it worthwhile to return both the integer cents and the decimal dollars? It feels a bit odd for the API to convert the list price to cents only for the caller of the API (learner portal) to convert it back to decimal dollars. A potential idea:

```js
{
  "list_price": {
    "USD": 299,
    "USD_CENTS": 29900,
  }
}
```

adamstankiewicz replied:
Related, locally, I'm seeing a course price of $299.00 in course-discovery be returned as `299000` in the `can-redeem` response, which would equate to `$29,900`.

It looks like `get_subsidy_content_data` is already returning integer cents?

pwnage101 replied:
I think I would agree about being explicit about the type, since the unit in this case doesn't directly pertain to any unit of any particular ledger/subsidy.

pwnage101 replied:
updated to match your proposed format.  Note that `usd` is a float, and `usd_cents` is an integer.

adamstankiewicz commented on enterprise_access/apps/api/serializers/subsidy_access_policy.py:
```diff
@@ -96,59 +96,52 @@ def validate(self, attrs):
 
 
-class ValidateContentKeyMixin:
+class ContentKeyField(serializers.CharField):
```
+1 this is a nice addition :)

adamstankiewicz commented on enterprise_access/apps/api/v1/views/subsidy_access_policy.py:
```diff
@@ -566,22 +591,43 @@ def can_redeem(self, request, enterprise_customer_uuid):
+                # Note that the "content_price" key is guaranteed to exist, but the value may be None.  Also, the
+                # upstream field type is a string representing a decimal dollar.
+                list_price_decimal_dollars_str = content_metadata["content_price"]
+                list_price_integer_cents = int(float(list_price_decimal_dollars_str) * 100)
```
If `list_price_decimal_dollars_str` can be `None`, won't an exception be raised here: `float(list_price_decimal_dollars_str)`?

```python
>>> float(None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: float() argument must be a string or a number, not 'NoneType'
```

The try/except doesn't catch `TypeError` though, so I believe this exception would be raised.

pwnage101 replied:
handled case, and added new test.

adamstankiewicz reviewed (approved):
LGTM!

iloveagent57 commented on enterprise_access/apps/api/v1/views/subsidy_access_policy.py:
```diff
@@ -248,6 +250,21 @@ class SubsidyAccessPolicyRedeemViewset(UserDetailsFromJwtMixin, PermissionRequir
+    @functools.lru_cache(maxsize=None)
+    def get_subsidy_client(cls):
+        """
+        A request-cached EnterpriseSubsidyAPIClient instance.
```
FWIW, I think `lru_cache` will cache for the lifetime of the process, not the request lifetime.

pwnage101 replied:
Interesting, that does seem to match up with my experience with these having lasting memory between unit tests, which in theory should destroy and create new instances of these views and models.  I'll update all the docstrings for now to say that they're process-cached since I think there's enough of a combination of sentiment and evidence to support the claim.  Note that `@cached_property` will have the same behavior since it's effectively backed by `lru_cache`.

iloveagent57 commented on enterprise_access/apps/api/v1/views/subsidy_access_policy.py:
```diff
@@ -626,22 +608,57 @@ def can_redeem(self, request, enterprise_customer_uuid):
             )
-            can_redeem_for_content_response = {
+
+            # Determine the price for content for display purposes only.
+            try:
+                content_metadata = self.subsidy_client.get_subsidy_content_data(enterprise_customer_uuid, content_key)
+                # Note that the "content_price" key is guaranteed to exist, but the value may be None.  Also, the
+                # upstream field type is a string representing a decimal dollar.
+                list_price_decimal_dollars_str = content_metadata["content_price"]
+                if list_price_decimal_dollars_str:
+                    list_price_decimal_dollars = float(list_price_decimal_dollars_str)
+                    list_price_integer_cents = int(list_price_decimal_dollars * 100)
+                else:
+                    list_price_decimal_dollars = None
+                    list_price_integer_cents = None
+            except requests.exceptions.HTTPError:
+                # No need to record a failure reason here because evaluate_policies() -> policy.can_redeem() already
+                # caught any problematic or non-existent content keys, and provided a reason it could not be redeemed.
+                # Just set a None content_price and move along.
+                list_price_decimal_dollars = None
+                list_price_integer_cents = None
```
This method body is getting really long - is there an obvious place to wrap this in a helper method or something?  Or maybe the serializer can do the translation of string to numbers?

pwnage101 replied:
I'd love to consolidate this into the serializer, but in light of my long todo list I'm going to defer this and add a comment.

iloveagent57 reviewed (approved):
Looks good, two small notes.
